File: lib/helper/conjugation/conj.py
import csv
import re



source = "//based on conj.csv which was scraped from\n//https://gitlab.com/yamagoya/jmdictdb/-/tree/master/jmdictdb/data\n\n\n\n"
base_path = "./lib/helper/conjugation/"
file_name = "conj"

def main():

    with open(f"{base_path}{file_name}.dart", mode="w+", encoding="utf8") as f:
        f.write(source)

        f.write("/// Enum containing the conjugation forms\n")
        f.write("enum Conj {\n")

        with open(f"{base_path}{file_name}.csv", newline='') as csvfile:

            conj_reader = csv.reader(csvfile, delimiter='\t', quotechar='|')

            for row in conj_reader:
                if(row[0] == "id"):
                    continue

                r = re.sub(r" \(.*\)", "", row[1]).replace("-", "_")
                f.write(f"\t{r},\n")
        
        f.write("}\n\n")

        f.write("/// A map from the id to the conjugation form enum\n")
        f.write("const Map<int, Conj> conjIdToConjEnum = {\n")

        with open(f"{base_path}{file_name}.csv", newline='') as csvfile:

            conj_reader = csv.reader(csvfile, delimiter='\t', quotechar='|')

            for row in conj_reader:
                if(row[0] == "id"):
                    continue

                r = re.sub(r" \(.*\)", "", row[1]).replace("-", "_")
                f.write(f"\t{row[0]} : Conj.{r},\n")
        
        f.write("};\n\n")


        f.write("/// A map from the conjugation form enum to the id\n")
        f.write("const Map<Conj, int> conjEnumToConjId = {\n")

        with open(f"{base_path}{file_name}.csv", newline='') as csvfile:

            conj_reader = csv.reader(csvfile, delimiter='\t', quotechar='|')

            for row in conj_reader:
                if(row[0] == "id"):
                    continue

                r = re.sub(r" \(.*\)", "", row[1]).replace("-", "_")
                f.write(f"\tConj.{r} : {row[0]},\n")
        
        f.write("};\n\n")

        f.write("/// A map from the conjugation form enum to a string description\n")
        f.write("const Map<Conj, String> conjEnumToConjDescription = {\n")

        with open(f"{base_path}{file_name}.csv", newline='') as csvfile:

            conj_reader = csv.reader(csvfile, delimiter='\t', quotechar='|')

            for row in conj_reader:
                if(row[0] == "id"):
                    continue

                r = re.sub(r" \(.*\)", "", row[1]).replace("-", "_")
                f.write(f"\tConj.{r} : '{row[1]}',\n")
        
        f.write("};")




if __name__ == "__main__":
    main()

File: lib/helper/conjugation/conj.dart
//based on conj.csv which was scraped from
//https://gitlab.com/yamagoya/jmdictdb/-/tree/master/jmdictdb/data



/// Enum containing the conjugation forms
enum Conj {
	Non_past,
	Past,
	Conjunctive,
	Provisional,
	Potential,
	Passive,
	Causative,
	Causative_Passive,
	Volitional,
	Imperative,
	Conditional,
	Alternative,
	Continuative,
}

/// A map from the id to the conjugation form enum
const Map<int, Conj> idToConj = {
	1 : Conj.Non_past,
	2 : Conj.Past,
	3 : Conj.Conjunctive,
	4 : Conj.Provisional,
	5 : Conj.Potential,
	6 : Conj.Passive,
	7 : Conj.Causative,
	8 : Conj.Causative_Passive,
	9 : Conj.Volitional,
	10 : Conj.Imperative,
	11 : Conj.Conditional,
	12 : Conj.Alternative,
	13 : Conj.Continuative,
};

/// A map from the conjugation form enum to the id
const Map<Conj, int> conjToId = {
	Conj.Non_past : 1,
	Conj.Past : 2,
	Conj.Conjunctive : 3,
	Conj.Provisional : 4,
	Conj.Potential : 5,
	Conj.Passive : 6,
	Conj.Causative : 7,
	Conj.Causative_Passive : 8,
	Conj.Volitional : 9,
	Conj.Imperative : 10,
	Conj.Conditional : 11,
	Conj.Alternative : 12,
	Conj.Continuative : 13,
};

/// A map from the conjugation form enum to a string description
const Map<Conj, String> conjToDescription = {
	Conj.Non_past : 'Non-past',
	Conj.Past : 'Past (~ta)',
	Conj.Conjunctive : 'Conjunctive (~te)',
	Conj.Provisional : 'Provisional (~eba)',
	Conj.Potential : 'Potential',
	Conj.Passive : 'Passive',
	Conj.Causative : 'Causative',
	Conj.Causative_Passive : 'Causative-Passive',
	Conj.Volitional : 'Volitional',
	Conj.Imperative : 'Imperative',
	Conj.Conditional : 'Conditional (~tara)',
	Conj.Alternative : 'Alternative (~tari)',
	Conj.Continuative : 'Continuative (~i)',
};